sdk/samples/iot/paho_iot_adu_sample_common.c

```diff
@@ -0,0 +1,663 @@
+    az_iot_hub_client_properties_message const* property_message)
+{
+  az_span const message_span = az_span_create((uint8_t*)message->payload, message->payloadlen);
+
+  // Invoke appropriate action per message type (3 types only).
+  switch (property_message->message_type)
+  {
+    // A message from a property GET publish message with the property document as a payload.
+    case AZ_IOT_HUB_CLIENT_PROPERTIES_MESSAGE_TYPE_GET_RESPONSE:
+      IOT_SAMPLE_LOG("Message Type: GET");
+      process_device_property_message(message_span, property_message->message_type);
+      break;
+
+    // An update to the desired properties with the properties as a payload.
+    case AZ_IOT_HUB_CLIENT_PROPERTIES_MESSAGE_TYPE_WRITABLE_UPDATED:
+      IOT_SAMPLE_LOG("Message Type: Desired Properties");
+      process_device_property_message(message_span, property_message->message_type);
+      break;
+
+    // When the device publishes a property update, this message type arrives when
+    // server acknowledges this.
+    case AZ_IOT_HUB_CLIENT_PROPERTIES_MESSAGE_TYPE_ACKNOWLEDGEMENT:
+      IOT_SAMPLE_LOG("Message Type: IoT Hub has acknowledged properties that the device sent");
+      break;
+
+    // An error has occurred
+    case AZ_IOT_HUB_CLIENT_PROPERTIES_MESSAGE_TYPE_ERROR:
+      IOT_SAMPLE_LOG_ERROR("Message Type: Request Error");
+      break;
+  }
+}
+
+// process_device_property_message handles incoming properties from Azure IoT Hub.
+static void process_device_property_message(
+    az_span message_span,
+    az_iot_hub_client_properties_message_type message_type)
+{
+  az_json_reader jr;
+  az_result rc = az_json_reader_init(&jr, message_span, NULL);
+  IOT_SAMPLE_EXIT_IF_AZ_FAILED(rc, "Could not initialize json reader");
+
+  int32_t version_number;
+  rc = az_iot_hub_client_properties_get_properties_version(
+      &hub_client, &jr, message_type, &version_number);
+  IOT_SAMPLE_EXIT_IF_AZ_FAILED(rc, "Could not get property version");
+
+  rc = az_json_reader_init(&jr, message_span, NULL);
+  IOT_SAMPLE_EXIT_IF_AZ_FAILED(rc, "Could not initialize json reader");
+
+  az_span component_name;
+
+  az_span xScratchBufferSpan
+      = az_span_create(adu_scratch_buffer, (int32_t)sizeof(adu_scratch_buffer));
+
+  // Applications call az_iot_hub_client_properties_get_next_component_property to enumerate
+  // properties received.
+  while (az_result_succeeded(az_iot_hub_client_properties_get_next_component_property(
+      &hub_client, &jr, message_type, AZ_IOT_HUB_CLIENT_PROPERTY_WRITABLE, &component_name)))
+  {
+    if (az_iot_adu_is_component_device_update(component_name))
+    {
+      // ADU Component
+      rc = az_iot_adu_parse_service_properties(
+          &jr, xScratchBufferSpan, &xBaseUpdateRequest, &xScratchBufferSpan);
+
+      if (az_result_failed(rc))
+      {
+        IOT_SAMPLE_LOG_ERROR("az_iot_adu_parse_service_properties failed: 0x%08x.", rc);
+        /* TODO: return individualized/specific errors. */
```

Comment:
Please address the TODOs - these should not be in or should be tracked by a GitHub issue.